In [1]:
import pandas as pd 
import json
import ast

# Here I will Merge the following csv files : 

* Base_data
* bbrinfobox
* buildingInfo
* unitInfo
* ownershipinfo
* lotInfo
* cadastralInfos
* additionlBuildingInfos
* floorInfos


In [2]:
############################################################################# base data
base_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\1_Base_data\Boliga_propertySales_Base_Data.csv'
data_base = pd.read_csv(base_path, encoding='utf-8', low_memory=False)
data_base.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 20 columns):
 #   Column                                        Non-Null Count    Dtype  
---  ------                                        --------------    -----  
 0   RowID_MAIN_boliga_ROW_ID_unitID               1031507 non-null  object 
 1   RowID_propertyCode_bfeCode                    1031507 non-null  int64  
 2   RowID_municipalityCode                        1031507 non-null  int64  
 3   RowID_propertyCode_esrCode                    1031507 non-null  int64  
 4   Address_info_addressString                    1031507 non-null  object 
 5   Address_info_streetString                     1031507 non-null  object 
 6   Address_info_houseNumber                      1031507 non-null  object 
 7   Address_info_floorString                      270077 non-null   object 
 8   Address_info_sideDoorNumber                   215286 non-null   object 
 9   Address_info_zipString_code_and_nam

In [3]:
############################################################################# 2_bbrInfoBox
bbrBoxInfo_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\2_bbrInfoBox\Ready\Boliga_propertySales_bbrInfoBox_READY.csv'
data_bbrBoxInfo = pd.read_csv(bbrBoxInfo_path, encoding='utf-8', low_memory=False)
data_bbrBoxInfo.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 8 columns):
 #   Column                           Non-Null Count    Dtype  
---  ------                           --------------    -----  
 0   RowID_MAIN_boliga_ROW_ID_unitID  1031507 non-null  object 
 1   bbrInfoBox_monthlyPropertyTax    1031507 non-null  float64
 2   bbrInfoBox_lotSize               1031507 non-null  int64  
 3   bbrInfoBox_roomsQuantity         1031507 non-null  int64  
 4   bbrInfoBox_area                  1031507 non-null  int64  
 5   bbrInfoBox_areaRecorded          1031507 non-null  int64  
 6   bbrInfoBox_evaluationPrice       1031507 non-null  int64  
 7   bbrInfoBox_basementSize          1031507 non-null  int64  
dtypes: float64(1), int64(6), object(1)
memory usage: 63.0+ MB


In [4]:
############################################################################# 3_buildingInfo
buildingInfo_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\3_buildingInfo\Ready\Boliga_propertySales_buildingInfo_READY.csv'
data_buildingInfo = pd.read_csv(buildingInfo_path, encoding='utf-8', low_memory=False)
data_buildingInfo.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 37 columns):
 #   Column                                           Non-Null Count    Dtype 
---  ------                                           --------------    ----- 
 0   RowID_MAIN_boliga_ROW_ID_unitID                  1031507 non-null  object
 1   buildingInfo_buildingNumber_BBR_lastdigit        1031507 non-null  int64 
 2   buildingInfo_BBR_buildingUsage_category          1031507 non-null  object
 3   buildingInfo_Year_built                          1031507 non-null  int64 
 4   buildingInfo_floorQuantity_in_building           1031507 non-null  int64 
 5   buildingInfo_latesstRebuildYear                  1031507 non-null  int64 
 6   buildingInfo_total_buildingArea                  1031507 non-null  int64 
 7   buildingInfo_total_businessArea_in_building      1031507 non-null  int64 
 8   buildingInfo_total_livingArea_in_building        1031507 non-null  int64 
 9   buildingInfo_

In [5]:
############################################################################# 4_unitInfo
unitInfo_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\4_unitInfo\Ready\Boliga_propertySales_unitInfo_READY.csv'
data_unitInfo = pd.read_csv(unitInfo_path, encoding='utf-8', low_memory=False)
data_unitInfo.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 24 columns):
 #   Column                                    Non-Null Count    Dtype 
---  ------                                    --------------    ----- 
 0   RowID_MAIN_boliga_ROW_ID_unitID           1031507 non-null  object
 1   unitInfo_appliance_usage                  1031507 non-null  object
 2   unitInfo_propertyUnitType                 1031507 non-null  object
 3   unitInfo_propertyCondemned                1031507 non-null  object
 4   unitInfo_Total_areaUnit                   1031507 non-null  int64 
 5   unitInfo_areaLiving                       1031507 non-null  int64 
 6   unitInfo_areaBusiness                     1031507 non-null  int64 
 7   unitInfo_areaOther                        1031507 non-null  int64 
 8   unitInfo_areaSocial_unitSharedCommonArea  1031507 non-null  int64 
 9   unitInfo_roomsQuantity                    1031507 non-null  int64 
 10  unitInfo_areaClose

In [6]:
############################################################################# 5_ownerShipInfo
ownerShipInfo_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\5_ownerShipInfo\Ready\Boliga_propertySales_ownerShipInfo_READY.csv'
data_ownerShipInfo= pd.read_csv(ownerShipInfo_path, encoding='utf-8', low_memory=False)
data_ownerShipInfo.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 4 columns):
 #   Column                                  Non-Null Count    Dtype 
---  ------                                  --------------    ----- 
 0   RowID_MAIN_boliga_ROW_ID_unitID         1031507 non-null  object
 1   ownerShipInfo_ownerShipCode             1031507 non-null  object
 2   ownerShipInfo_swimmingPool_dummy        1031507 non-null  int64 
 3   ownershipInfo_greehouse_dummy_variable  1031507 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 31.5+ MB


In [7]:
############################################################################# 6_lotInfo
lotInfo_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\6_lotInfo\Ready\Boliga_propertySales_lotInfo.csv'
data_lotInfo = pd.read_csv(lotInfo_path, encoding='utf-8', low_memory=False)
data_lotInfo.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 4 columns):
 #   Column                           Non-Null Count    Dtype 
---  ------                           --------------    ----- 
 0   RowID_MAIN_boliga_ROW_ID_unitID  1031507 non-null  object
 1   lotInfo_watersupplyCode          853707 non-null   object
 2   lotInfo_swimmingPool_dummy       1031507 non-null  int64 
 3   lotInfo_greehouse_dummy          1031507 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 31.5+ MB


In [8]:
############################################################################# 7_cadastral
cadastralInfo_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\7_cadastralInfos\Ready\Boliga_propertySales_CadastralInfo.csv'
data_cadastralInfo = pd.read_csv(cadastralInfo_path, encoding='utf-8', low_memory=False)
data_cadastralInfo.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 3 columns):
 #   Column                                                  Non-Null Count    Dtype 
---  ------                                                  --------------    ----- 
 0   RowID_MAIN_boliga_ROW_ID_unitID                         1031507 non-null  object
 1   cadestalsInfos_Number_of_cadestalsInfos_dictionaries    1031507 non-null  int64 
 2   cadestalsInfos_Total_land_size_with_property_grundArea  1031507 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 23.6+ MB


In [9]:
############################################################################# 8_additinalBuildingInfos
additinalBuildingInfos_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\8_additinalBuildingInfos\Ready\Boliga_propertySales_AdditionalBuildingInfos.csv'
data_additinalBuildingInfos = pd.read_csv(additinalBuildingInfos_path, encoding='utf-8', low_memory=False)
data_additinalBuildingInfos.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 29 columns):
 #   Column                                                            Non-Null Count    Dtype  
---  ------                                                            --------------    -----  
 0   RowID_MAIN_boliga_ROW_ID_unitID                                   1031507 non-null  object 
 1   additionarBuildingInfo_Carport_Exists_dummy                       1031507 non-null  float64
 2   additionarBuildingInfo_numberOf_Carports_on_property              1031507 non-null  float64
 3   additionarBuildingInfo_Carport_Total_size_sumedUp                 1031507 non-null  float64
 4   additionarBuildingInfo_Outhouse_Exists_dummy                      1031507 non-null  float64
 5   additionarBuildingInfo_numberOf_Outhouse_on_property              1031507 non-null  float64
 6   additionarBuildingInfo_Outhouse_Total_size_sumedUp                1031507 non-null  float64
 7   additiona

In [10]:
############################################################################# 9_floorInfos
floorInfos_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_split\9_floorInfos\Ready\Boliga_propertySales_floorInfos_Ready.csv'
data_floorInfos = pd.read_csv(floorInfos_path, encoding='utf-8', low_memory=False)
data_floorInfos.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 4 columns):
 #   Column                           Non-Null Count    Dtype  
---  ------                           --------------    -----  
 0   RowID_MAIN_boliga_ROW_ID_unitID  1031507 non-null  object 
 1   floorInfo_Basement_Exists_dummy  1031507 non-null  float64
 2   floorInfo_Basement_totalarea     1031507 non-null  float64
 3   floorInfo_Basement_livingarea    1031507 non-null  float64
dtypes: float64(3), object(1)
memory usage: 31.5+ MB


# ----------------------- MERGE DATA -------------------------

In [12]:
# data_base + data_bbrBoxInfo
data_MERGE = pd.merge(data_base, data_bbrBoxInfo, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

# data_MERGE +  data_buildingInfo
data_MERGE = pd.merge(data_MERGE, data_buildingInfo, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

# data_MERGE +  unitinfo
data_MERGE = pd.merge(data_MERGE, data_unitInfo, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

# data_MERGE +  ownershipinfo
data_MERGE = pd.merge(data_MERGE, data_ownerShipInfo, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

# data_MERGE +  lotInfo
data_MERGE = pd.merge(data_MERGE, data_lotInfo, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

# data_MERGE +  cadastral infos
data_MERGE = pd.merge(data_MERGE, data_cadastralInfo, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

# data_MERGE + data_additionalBuildingInfos
data_MERGE = pd.merge(data_MERGE, data_additinalBuildingInfos, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

# data_MERGE +  floorinfos
data_MERGE = pd.merge(data_MERGE, data_floorInfos, on= 'RowID_MAIN_boliga_ROW_ID_unitID', how='left')

data_MERGE.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 125 columns):
 #    Column                                                            Non-Null Count    Dtype  
---   ------                                                            --------------    -----  
 0    RowID_MAIN_boliga_ROW_ID_unitID                                   1031507 non-null  object 
 1    RowID_propertyCode_bfeCode                                        1031507 non-null  int64  
 2    RowID_municipalityCode                                            1031507 non-null  int64  
 3    RowID_propertyCode_esrCode                                        1031507 non-null  int64  
 4    Address_info_addressString                                        1031507 non-null  object 
 5    Address_info_streetString                                         1031507 non-null  object 
 6    Address_info_houseNumber                                          1031507 non-null  object 
 7  

In [13]:
############## rename 2 columns 
rename_these = {
    'RowID_propertyCode_bfeCode' : 'RowID_propertyCode_BFE_CODE',                                         
    'RowID_municipalityCode' : 'RowID_municipalityCode_CODE'                                                                               
}
data_MERGE.rename(columns=rename_these, inplace=True)

In [14]:
######################## delete unnecessary frames taking space 
del base_path, data_base, bbrBoxInfo_path, data_bbrBoxInfo, buildingInfo_path, data_buildingInfo, unitInfo_path, data_unitInfo, ownerShipInfo_path, data_ownerShipInfo, lotInfo_path, data_lotInfo, cadastralInfo_path, data_cadastralInfo, additinalBuildingInfos_path, data_additinalBuildingInfos, floorInfos_path, data_floorInfos 

import gc
gc.collect()

169

# ------------------ GET THE SFE property and municipality codes 

( SFE codes are for properties with shared estates e.g. apartments !!!!)

### Right now I only have property and municipality BFE Codes - some properties did not have energy reports based on their BFE codes but did have energy reports on SFE codes e.g. building with multiple apartments 
### therefore I need those codes to create the MERGING property code and Merging municipality codes  columnes to merge with the EPC data 

In [15]:
############################################################################# Get EPC data
path_SFE = r'D:\Thesis\EPC\EPC_PDF_from_BoligaDK_data\1_Download_PDF_files\Property_URLs_of_pdf_files\step3_Scrape_apartments_based_on_SFE_number\Beautifulsoup_Boliga_apartments_EPC_urls.csv'
data_SFE = pd.read_csv(path_SFE, encoding='utf-8', low_memory=False)
data_SFE.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 276542 entries, 0 to 276541
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   guid_code              276542 non-null  object 
 1   property_esrCode_BFE   276542 non-null  int64  
 2   municipalityCode_BFE   276542 non-null  int64  
 3   addressString          276542 non-null  object 
 4   streetname             276542 non-null  object 
 5   address_number         276542 non-null  object 
 6   Zip_code_name          276542 non-null  object 
 7   Zip_code               276542 non-null  int64  
 8   isApartment            276542 non-null  bool   
 9   bfenr                  276542 non-null  int64  
 10  url_ENERGI             276542 non-null  object 
 11  municipality_code_SFE  275906 non-null  float64
 12  property_code_SFE      275906 non-null  float64
 13  http_response          275906 non-null  float64
 14  epc_exists_dummy       276542 non-nu

In [16]:
############################################################################# Get EPC data
path_SFE2 = r'D:\Thesis\EPC\EPC_PDF_from_BoligaDK_data\1_Download_PDF_files\Property_URLs_of_pdf_files\step9_FINAL_check_which_property_EPC_reports_are_missing_compared_to_old_address_list\Boliga_Addresslist_Final.csv'
data_SFE2 = pd.read_csv(path_SFE2, encoding='utf-8', low_memory=False)
data_SFE2.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1606925 entries, 0 to 1606924
Data columns (total 18 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   Unnamed: 0              1606925 non-null  int64  
 1   guid_code               1606925 non-null  object 
 2   property_esrCode_BFE    1606925 non-null  int64  
 3   municipalityCode_BFE    1606925 non-null  int64  
 4   addressString           1606925 non-null  object 
 5   streetname              1606925 non-null  object 
 6   address_number          1606925 non-null  object 
 7   Zip_code_name           1606925 non-null  object 
 8   Zip_code                1606925 non-null  int64  
 9   isApartment             1606925 non-null  bool   
 10  bfenr                   1606925 non-null  int64  
 11  http_response           1606271 non-null  float64
 12  epc_exists_dummy        1606925 non-null  float64
 13  pdf_URL                 1457007 non-null  object 
 14  pd

In [19]:
data_SFE2['bfenr']

0           2288926
1          10085484
3           3375081
4           5432373
5           2173061
             ...   
1606917     4246756
1606919     7488964
1606921     7488964
1606923     6009990
1606924     2530339
Name: bfenr, Length: 981393, dtype: int64

In [17]:
###################### drop duplicated guid_codes from boliga
data_SFE2 = data_SFE2.drop_duplicates(subset=['guid_code'])
print(len(data_SFE2))

981393


In [20]:
################### filter the SFE data for only the guid_code and municipality_code_SFE  and property_code_SFE      
columns_to_keep = ['guid_code', 'municipality_code_SFE','property_code_SFE']
data_SFE2 = data_SFE2[columns_to_keep]


################### Rename the columns       
rename_columns = {
    'guid_code' : 'RowID_MAIN_boliga_ROW_ID_unitID', 
    'municipality_code_SFE' : 'RowID_municipalityCode_SFE_CODE',
    'property_code_SFE' : 'RowID_propertyCode_SFE_CODE'
}

data_SFE2.rename(columns=rename_columns, inplace=True)

data_SFE2.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
Index: 981393 entries, 0 to 1606924
Data columns (total 3 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   RowID_MAIN_boliga_ROW_ID_unitID  981393 non-null  object 
 1   RowID_municipalityCode_SFE_CODE  176039 non-null  float64
 2   RowID_propertyCode_SFE_CODE      176039 non-null  float64
dtypes: float64(2), object(1)
memory usage: 29.9+ MB


## ------------ MERGE data_MERGE and data_SFE

In [21]:
data_MERGE = pd.merge(data_MERGE, data_SFE2, on='RowID_MAIN_boliga_ROW_ID_unitID', how='left')
data_MERGE.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 127 columns):
 #    Column                                                            Non-Null Count    Dtype  
---   ------                                                            --------------    -----  
 0    RowID_MAIN_boliga_ROW_ID_unitID                                   1031507 non-null  object 
 1    RowID_propertyCode_BFE_CODE                                       1031507 non-null  int64  
 2    RowID_municipalityCode_CODE                                       1031507 non-null  int64  
 3    RowID_propertyCode_esrCode                                        1031507 non-null  int64  
 4    Address_info_addressString                                        1031507 non-null  object 
 5    Address_info_streetString                                         1031507 non-null  object 
 6    Address_info_houseNumber                                          1031507 non-null  object 
 7  

In [23]:
data_MERGE['RowID_propertyCode_SFE_CODE'].unique()

array([    nan, 688453., 335250., ..., 277138., 140833.,  17388.])

In [24]:
# data_MERGE = pd.merge(data_MERGE, data_SFE2, on='RowID_MAIN_boliga_ROW_ID_unitID', how='left')
# data_MERGE.info(verbose=True, show_counts=True)

In [25]:
del path_SFE, data_SFE

import gc
gc.collect()

0

In [26]:
data_MERGE['RowID_propertyCode_esrCode'].unique()

array([981498, 134998,   4216, ...,  72508,  60576, 451454], dtype=int64)

In [27]:
data_MERGE['RowID_propertyCode_BFE_CODE'].unique()

array([10085484,  2173061,  5432373, ...,  5505660,  4334642,  2623035],
      dtype=int64)

# ---- Create MERGIN COLUMNS with EPC data in the sales dataframe : 

## the columns are Merge_property_ID_CODE and Merge_municipality_ID_CODE

I simply create a copy of the '**RowID_municipalityCode_SFE_CODE**' and '**RowID_propertyCode_SFE_CODE**'  THEN fill na values with '**RowID_propertyCode_BFE_CODE**' and  '**RowID_municipalityCode_BFE_CODE**'                                                                                                 

In [28]:
############## make the copy
data_MERGE['Merge_property_ID_CODE'] = data_MERGE['RowID_propertyCode_SFE_CODE']
data_MERGE['Merge_municipality_ID_CODE'] =  data_MERGE['RowID_municipalityCode_SFE_CODE']

############## fill na 
data_MERGE['Merge_property_ID_CODE'] = data_MERGE['Merge_property_ID_CODE'].fillna(data_MERGE['RowID_propertyCode_esrCode']) 
data_MERGE['Merge_municipality_ID_CODE'] = data_MERGE['Merge_municipality_ID_CODE'].fillna(data_MERGE['RowID_municipalityCode_CODE']) 


In [29]:
data_MERGE.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031507 entries, 0 to 1031506
Data columns (total 129 columns):
 #    Column                                                            Non-Null Count    Dtype  
---   ------                                                            --------------    -----  
 0    RowID_MAIN_boliga_ROW_ID_unitID                                   1031507 non-null  object 
 1    RowID_propertyCode_BFE_CODE                                       1031507 non-null  int64  
 2    RowID_municipalityCode_CODE                                       1031507 non-null  int64  
 3    RowID_propertyCode_esrCode                                        1031507 non-null  int64  
 4    Address_info_addressString                                        1031507 non-null  object 
 5    Address_info_streetString                                         1031507 non-null  object 
 6    Address_info_houseNumber                                          1031507 non-null  object 
 7  

.

# ------------------------------------------------------------

# READ IN THE NEW ready Sales data !

# ------------------------------------------------------------

In [30]:
############################################################################# Get EPC data
SALES_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_merge\step_2_mrgin_newSales_with_Estate_ads\Boliga_salesData_SalesTransactions_with_evaluation_and_estate_info_34col_2623340rows.csv'

SALES_data = pd.read_csv(SALES_path, encoding='utf-8', low_memory=False)
SALES_data.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2623340 entries, 0 to 2623339
Data columns (total 40 columns):
 #   Column                                           Non-Null Count    Dtype  
---  ------                                           --------------    -----  
 0   RowID_MAIN_boliga_ROW_ID_unitID                  2623340 non-null  object 
 1   salesInfo_isSalesValid                           2623340 non-null  bool   
 2   salesInfo_handoverCode                           2623340 non-null  float64
 3   salesInfo_handoverName                           2623340 non-null  object 
 4   salesInfo_deedIssueDate                          2623340 non-null  object 
 5   salesInfo_price                                  2623340 non-null  float64
 6   salesInfo_recalculationDate                      2623340 non-null  object 
 7   salesInfo_rebuildYear                            2623340 non-null  float64
 8   unique_sales_ID                                  2623340 non-null  int64  
 9   sa

.

# ------------------------------------------------------------

# MERGE Sales and base data 
final data from this should have 2623340  rows !

# ------------------------------------------------------------

In [31]:
data_MERGE = pd.merge(SALES_data, data_MERGE, on='RowID_MAIN_boliga_ROW_ID_unitID', how='left')
data_MERGE.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2623340 entries, 0 to 2623339
Data columns (total 168 columns):
 #    Column                                                            Non-Null Count    Dtype  
---   ------                                                            --------------    -----  
 0    RowID_MAIN_boliga_ROW_ID_unitID                                   2623340 non-null  object 
 1    salesInfo_isSalesValid                                            2623340 non-null  bool   
 2    salesInfo_handoverCode                                            2623340 non-null  float64
 3    salesInfo_handoverName                                            2623340 non-null  object 
 4    salesInfo_deedIssueDate                                           2623340 non-null  object 
 5    salesInfo_price                                                   2623340 non-null  float64
 6    salesInfo_recalculationDate                                       2623340 non-null  object 
 7  

In [34]:
columns_to_Drop = [
    'Estate_Info_NEWESTAd_askingPrice_lastAd',
    'Estate_Info_NEWESTAd_daysForSale',
    'Estate_Info_NEWESTAd_lastSeen_Date',
    'Estate_Info_OLDESTAd_askingPrice_firstAd',
    'Estate_Info_OLDESTAd_daysForSale',
    'Estate_Info_OLDESTAd_lastSeen_Date',
]
data_MERGE.drop(columns=columns_to_Drop, inplace=True)
data_MERGE.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2623340 entries, 0 to 2623339
Data columns (total 162 columns):
 #    Column                                                            Non-Null Count    Dtype  
---   ------                                                            --------------    -----  
 0    RowID_MAIN_boliga_ROW_ID_unitID                                   2623340 non-null  object 
 1    salesInfo_isSalesValid                                            2623340 non-null  bool   
 2    salesInfo_handoverCode                                            2623340 non-null  float64
 3    salesInfo_handoverName                                            2623340 non-null  object 
 4    salesInfo_deedIssueDate                                           2623340 non-null  object 
 5    salesInfo_price                                                   2623340 non-null  float64
 6    salesInfo_recalculationDate                                       2623340 non-null  object 
 7  

#### Result from the old merge :

In [22]:
data_MERGE = pd.merge(SALES_data, data_MERGE, on='RowID_MAIN_boliga_ROW_ID_unitID', how='left')
data_MERGE.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2623340 entries, 0 to 2623339
Data columns (total 162 columns):
 #    Column                                                            Non-Null Count    Dtype  
---   ------                                                            --------------    -----  
 0    RowID_MAIN_boliga_ROW_ID_unitID                                   2623340 non-null  object 
 1    salesInfo_isSalesValid                                            2623340 non-null  bool   
 2    salesInfo_handoverCode                                            2623340 non-null  float64
 3    salesInfo_handoverName                                            2623340 non-null  object 
 4    salesInfo_deedIssueDate                                           2623340 non-null  object 
 5    salesInfo_price                                                   2623340 non-null  float64
 6    salesInfo_recalculationDate                                       2623340 non-null  object 
 7  

In [ ]:
data_MERGE['RA']

In [33]:
################## clean memory from useless data

del SALES_path, SALES_data

# garbage collect 
import gc
gc.collect()

0

In [32]:
len(data_MERGE['unique_sales_ID'].isna())

2623340

# ------ SAVE RESULT 

In [35]:
save_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_merge\step_FINAL_Data_Ready\Main_Boliga_Sales_data_with_valuation_and_estates.csv'

data_MERGE.to_csv(save_path, encoding='utf-8', index=False)

.

# ------------------------------------------------------------

# create NEW dataframe of the Sales data to merge with EPC 

NOTE!: this will only include following columns:
* RowID_MAIN_boliga_ROW_ID_unitID (not unique key)
* salesInfo_deedIssueDate (to match with epc where epc report date must be older)
* salesInfo_recalculationDate (just in case...)
* unique_sales_ID (unique key for sales row)
* RowID_propertyCode_BFE_CODE 
* RowID_municipalityCode_BFE_CODE 
* RowID_municipalityCode_SFE_CODE
* RowID_propertyCode_SFE_CODE
* Merge_property_ID_CODE
* Merge_municipality_ID_CODE


# ------------------------------------------------------------

In [ ]:
'RowID_MAIN_boliga_ROW_ID_unitID',                                   
'salesInfo_deedIssueDate',                                           
'salesInfo_MAIN_date_of_Sale', 
'salesInfo_recalculationDate',                                       
'unique_sales_ID',                                                   
'RowID_propertyCode_BFE_CODE',                                       
'RowID_municipalityCode_CODE',                                       
'RowID_propertyCode_esrCode',                                        
'RowID_municipalityCode_SFE_CODE',                                   
'RowID_propertyCode_SFE_CODE',                                       
'Merge_property_ID_CODE',                                            
'Merge_municipality_ID_CODE',                                        

In [36]:
sub_column = [
    'RowID_MAIN_boliga_ROW_ID_unitID',                                   
    'salesInfo_deedIssueDate',                                           
    'salesInfo_MAIN_date_of_Sale', 
    'salesInfo_recalculationDate',                                       
    'unique_sales_ID',                                                   
    'RowID_propertyCode_BFE_CODE',                                       
    'RowID_municipalityCode_CODE',                                       
    'RowID_propertyCode_esrCode',                                        
    'RowID_municipalityCode_SFE_CODE',                                   
    'RowID_propertyCode_SFE_CODE',                                       
    'Merge_property_ID_CODE',                                            
    'Merge_municipality_ID_CODE',                                        
]
new_save_path = r'D:\Thesis\Properties\Denmark\RE_due_scraping_properties\Boliga_dk\Creating_main_dataset_for_sales_data\Data_merge\step_FINAL_Data_Ready\Boliga_Main_ID_to_merge_with_EPC.csv'
data_MERGE[sub_column].to_csv(new_save_path, encoding='utf-8', index=False)

In [35]:

data_MERGE[sub_column].info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2623340 entries, 0 to 2623339
Data columns (total 10 columns):
 #   Column                           Non-Null Count    Dtype  
---  ------                           --------------    -----  
 0   RowID_MAIN_boliga_ROW_ID_unitID  2623340 non-null  object 
 1   salesInfo_deedIssueDate          2623340 non-null  object 
 2   salesInfo_recalculationDate      2623340 non-null  object 
 3   unique_sales_ID                  2623340 non-null  int64  
 4   RowID_propertyCode_BFE_CODE      2623340 non-null  int64  
 5   RowID_municipalityCode_BFE_CODE  2623340 non-null  int64  
 6   RowID_municipalityCode_SFE_CODE  325992 non-null   float64
 7   RowID_propertyCode_SFE_CODE      325992 non-null   float64
 8   Merge_property_ID_CODE           2623340 non-null  float64
 9   Merge_municipality_ID_CODE       2623340 non-null  float64
dtypes: float64(4), int64(3), object(3)
memory usage: 200.1+ MB
